### PR TITLE
Add label support

### PR DESCRIPTION
**Pull request by HexRabbit**
Create a label by using "WORD:", and use ":WORD" to call/jump to that label.

Also modified all call/jmp-like instruction in test/*.s to match new usage of `$` and `#`
( `$` for CONST, `#` for TEMP )

### CODE REVIEW

**Comment by jserv** [comment]
`README.md` should be updated and revised accordingly. Also, all jump instructions are only followed by labels or valid "addresses".

**Comment by HexRabbit** [comment]
`README.md` will be updated later.
Using instant value or passing a temp is the same as passing a pointer to "address".
Old design uses `#X` to indicate address X (in jmp-like instructions), 
I think it's bad to have two meaning for `#`.

**Comment by jserv** [comment]
@HexRabbit, let's clarify these notations.

**Comment by HexRabbit** [comment]
I thought that `#` is a storage can store a value, and `$` is constant value,
but in jmp-like instructions, old design uses `#X` to indicate address X, 
make `#` have two meanings.
Did I misunderstand the meaning of these two?

**Comment by jserv** [comment]
Take Arm assembly for example. `#` is followed by immediate value (or simply an immediate or imm) while `$` is followed by mnemonic registers. @HexRabbit, feel free to submit another pull request to unify the usage in advance.

Reference: https://cs107e.github.io/lectures/ASM/asm.pdf

**Comment by HexRabbit** [comment]
I'm not familiar with Arm assembly(will take a look after), 
I just look into the code to understand what it means.
then I found this two, shows that its original design is not as same as Arm's.
So...what should I do, keep it the same or turning to Arm's?
```
"Supported types:\n"
              "       $ (constant integer)\n"
              "       # (temp integer)\n"
```
```  
  case '$':
        op.type = CONST;
        int temp = atoi(data + 1);  // filter first char
        op.value.id = vm_add_const(env, INT, &temp);
        break;
    case '#':
        op.type = TEMP;
        op.value.id = atoi(data + 1);
        break;
```

**Comment by HexRabbit** [comment]
Oh, I overlooked the code and just judging the meaning only by the word.
Sorry, I will revise the code.

**Comment by jserv** [comment]
In early stage of this project, I was designing stack-based VM rather than register-based one. While shifting to register-based design, there were various inconsistent decisions like what @HexRabbit mentioned in previous code snip. It would be great to clarify and refactor.

Cc. @pakls

**Comment by HexRabbit** [comment]
I just got confused by op.value.id. Sorry.
So, I will post a pull request for notation refactor first.

**Comment by jserv** [comment]
`README.md` should be update accordingly.

**Comment by HexRabbit** [comment]
Is there anything needs to be changed before merge?

**Comment by HexRabbit** [comment]
Updated `README.md`.

**Comment by jserv** [comment]
Always rebase to `master` branch, otherwise you will get commits like `Merge branch 'jmp-revise'`.

File: as.c [hunk 1]
```diff
@@ -177,7 +177,7 @@ static inline vm_operand make_operand(vm_env *env, char *line, const char *data)
     switch (data[0]) {
     case '$':
         op.type = CONST;
-        int temp = atoi(data + 1);
+        int temp = atoi(data + 1);  // filter first char
```
Comment (jserv) [inline as.c]: Use C-style comments rather than C++.

File: vm.c [hunk 2]
```diff
@@ -129,9 +134,41 @@ vm_env *vm_new()
 
 void vm_free(vm_env *env)
 {
+    for (vm_label **i = env->labels; i < env->labels + LABEL_HASH_TABLE_SIZE;
+         ++i) {
+        vm_label *tmp = *i;
+        while (tmp != NULL) {
```
Comment (jserv) [inline vm.c]: Rewrite as `while (tmp) { ... }`.

File: vm.c [hunk 3]
```diff
@@ -95,6 +96,9 @@ static inline void vm_push(vm_env *env, size_t n);
 /* OPCODE impl max size */
 #define OPCODE_IMPL_MAX_SIZE 256
 
+/* Label hash table size */
+#define LABEL_HASH_TABLE_SIZE 49157
```
Comment (jserv) [inline vm.c]: Rename to `LABEL_HT_SIZE`, which stands for the same but shorter.

File: as.c [hunk 4]
```diff
@@ -189,6 +189,18 @@ static inline vm_operand make_operand(vm_env *env, char *line, const char *data)
         op.type = CONST;
         op.value.id = vm_add_const(env, STR, quoted_strdup(data));
         break;
+    case ':':;
```
Comment (jserv) [inline as.c]: Why  `case ':':;` here?

File: vm.c (hunk 3 above)
Comment (HexRabbit) [reply vm.c]: Got it.

File: as.c (hunk 4 above)
Comment (HexRabbit) [reply as.c]: It seems that I cannot declare an variable in switch statement,
and that's one of the solution I found on stackoverflow.

And the another way doing this is by adding  a `{}` to make it a compound statement.

Comment (HexRabbit) [reply as.c]: After some researching on K&R C, 
I found that using `;` is just to satisfy the requirement in _labeled-statement_ -- a _null statement_
So `{}` will be the proper way for doing this.

**Comment by pakls** [comment]
A non-technical question: shall changes in this repo be pulled back to original repo?

**Comment by jserv** [comment]
@pakls, [sysprog21/full-stack-hello](https://github.com/sysprog21/full-stack-hello/) is meant to the placeholder of collaborative work from students while the origin accepts extra PRs.

**Comment by pakls** [comment]
Ok, after the semester ends, the patches of full-stack-hello could be
merged to the origin.

**Pull request merged**